les/serverpool.go

```diff
@@ -17,904 +17,283 @@
-			case pool.timeout <- entry:
-			}
-		}
-	}()
+	s.ns.UpdateState(p.ID(), s.stRedialWait, s.stConnected, time.Second*10)
+	s.vt.Unregister(p.ID())
+	p.setValueTracker(nil, nil)
 }
 
-// checkDialTimeout checks if the node is still in dialed state and if so, resets it
-// and adjusts connection statistics accordingly.
-func (pool *serverPool) checkDialTimeout(entry *poolEntry) {
-	if entry.state != psDialed {
-		return
+// getTimeout calculates the current recommended timeout. This value is used by
+// the client as a "soft timeout" value. It also affects the service value calculation
+// of individual nodes.
+func (s *serverPool) getTimeout() time.Duration {
+	now := s.clock.Now()
+	s.timeoutLock.RLock()
+	timeout := s.timeout
+	refreshed := s.timeoutRefreshed
+	s.timeoutLock.RUnlock()
+	if refreshed != 0 && time.Duration(now-refreshed) < timeoutRefresh {
+		return timeout
 	}
-	log.Debug("Dial timeout", "lesaddr", entry.node.ID().String()+"@"+entry.dialed.strKey())
-	entry.state = psNotConnected
-	if entry.knownSelected {
-		pool.knownSelected--
-	} else {
-		pool.newSelected--
+	rts := s.vt.RtStats()
+	rts.Add(time.Second*2, 10, s.vt.StatsExpFactor())
+	timeout = minTimeout
+	if t := rts.Timeout(0.1); t > timeout {
+		timeout = t
 	}
-	entry.connectStats.add(0, 1)
-	entry.dialed.fails++
-	pool.setRetryDial(entry)
-}
-
-const (
-	psNotConnected = iota
-	psDialed
-	psConnected
-	psRegistered
-)
-
-// poolEntry represents a server node and stores its current state and statistics.
-type poolEntry struct {
-	peer                  *serverPeer
-	pubkey                [64]byte // secp256k1 key of the node
-	addr                  map[string]*poolEntryAddress
-	node                  *enode.Node
-	lastConnected, dialed *poolEntryAddress
-	addrSelect            utils.WeightedRandomSelect
-
-	lastDiscovered                mclock.AbsTime
-	known, knownSelected, trusted bool
-	connectStats, delayStats      poolStats
-	responseStats, timeoutStats   poolStats
-	state                         int
-	regTime                       mclock.AbsTime
-	queueIdx                      int
-	removed                       bool
-
-	delayedRetry bool
-	shortRetry   int
-}
-
-// poolEntryEnc is the RLP encoding of poolEntry.
-type poolEntryEnc struct {
-	Pubkey                     []byte
-	IP                         net.IP
-	Port                       uint16
-	Fails                      uint
-	CStat, DStat, RStat, TStat poolStats
-}
-
-func (e *poolEntry) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, &poolEntryEnc{
-		Pubkey: encodePubkey64(e.node.Pubkey()),
-		IP:     e.lastConnected.ip,
-		Port:   e.lastConnected.port,
-		Fails:  e.lastConnected.fails,
-		CStat:  e.connectStats,
-		DStat:  e.delayStats,
-		RStat:  e.responseStats,
-		TStat:  e.timeoutStats,
-	})
-}
-
-func (e *poolEntry) DecodeRLP(s *rlp.Stream) error {
-	var entry poolEntryEnc
-	if err := s.Decode(&entry); err != nil {
-		return err
+	if t := rts.Timeout(0.5) * 2; t > timeout {
+		timeout = t
```

Comment:
I can't understand the logic here. 10 percent failure ratio seems conservative, 50 percent failure ratio seems aggressive. 

The logic here:
* If the conservative `to` is even smaller than minimal one, use minimal; otherwise record conservative `to`.
* If the aggressive `to` is larger than `conservative` to, then use aggressive one.

Reply:
`rts.Timeout(0.5)` is multiplied by 2, this is why it can be larger than the conservative timeout (otherwise this would not be possible). So we either use the 10% failure rate timeout or twice the median response time, whichever is more conservative. We can't really predict what kind of distribution we are going to have (might also depend on the connection type) so I am mostly making a guess at this point but I think twice the median is usually not too high and hopefully it can produce even less than 10% failure rate if the response times are predictable and the distribution is narrow. We do check it against the 10% failure threshold though so we are not going to have anything worse than that over the long term even if there are lots of slow responses and we have a "long tail" distribution.